src/main/kotlin/com/github/jengelman/gradle/plugins/shadow/transformers/ServicesResourceTransformer.kt

```diff
@@ -0,0 +1,97 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.maven.plugins.shade.resource
+
+import com.github.jengelman.gradle.plugins.shadow.relocation.RelocateClassContext
+import com.github.jengelman.gradle.plugins.shadow.transformers.Transformer
+import com.github.jengelman.gradle.plugins.shadow.transformers.TransformerContext
+import java.nio.charset.StandardCharsets
+import java.util.Scanner
+import org.apache.commons.io.IOUtils
+import org.apache.tools.zip.ZipEntry
+import org.apache.tools.zip.ZipOutputStream
+import org.gradle.api.file.FileTreeElement
+
+/**
+ * Resources transformer that relocates classes in META-INF/services and appends entries in META-INF/services resources
+ * into a single resource. For example, if there are several META-INF/services/org.apache.maven.project.ProjectBuilder
+ * resources spread across many JARs the individual entries will all be concatenated into a single
+ * META-INF/services/org.apache.maven.project.ProjectBuilder resource packaged into the resultant JAR produced by the
+ * shading process.
+ */
+private class ServicesResourceTransformer : Transformer {
+  private val serviceEntries: MutableMap<String, MutableSet<String?>> = HashMap()
```

Comment:
This property gets rid of `ServiceStream` in https://github.com/apache/maven-shade-plugin/commit/c18d3d3e36f4d37aba98a18983bd4e56eba79327.